From: Jason Dillaman <dillaman@redhat.com>
Date: Fri, 27 May 2016 11:40:58 -0400
Subject: rbd-mirror: cluster-level asok commands need to support multiple
 pools

Signed-off-by: Jason Dillaman <dillaman@redhat.com>
(cherry picked from commit f4339ee8cb64d5f1a164e017fea942600ffb0ea9)
(cherry picked from commit 0bcc295363d9ca20d43fd313084370158513bc62)

Resolves: rhbz#1340476
---
 src/tools/rbd_mirror/Replayer.cc | 14 ++++++++++++--
 src/tools/rbd_mirror/Replayer.h  |  2 ++
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/src/tools/rbd_mirror/Replayer.cc b/src/tools/rbd_mirror/Replayer.cc
index f3d4b03..dc6acf9 100644
--- a/src/tools/rbd_mirror/Replayer.cc
+++ b/src/tools/rbd_mirror/Replayer.cc
@@ -242,8 +242,6 @@ Replayer::Replayer(Threads *threads, std::shared_ptr<ImageDeleter> image_deleter
   m_asok_hook(nullptr),
   m_replayer_thread(this)
 {
-  CephContext *cct = static_cast<CephContext *>(m_local->cct());
-  m_asok_hook = new ReplayerAdminSocketHook(cct, m_peer.cluster_name, this);
 }
 
 Replayer::~Replayer()
@@ -408,6 +406,17 @@ void Replayer::run()
   dout(20) << "enter" << dendl;
 
   while (!m_stopping.read()) {
+
+    std::string asok_hook_name = m_local_io_ctx.get_pool_name() + " " +
+                                 m_peer.cluster_name;
+    if (m_asok_hook_name != asok_hook_name || m_asok_hook == nullptr) {
+      m_asok_hook_name = asok_hook_name;
+      delete m_asok_hook;
+
+      CephContext *cct = static_cast<CephContext *>(m_local->cct());
+      m_asok_hook = new ReplayerAdminSocketHook(cct, m_asok_hook_name, this);
+    }
+
     Mutex::Locker l(m_lock);
     if (!m_manual_stop) {
       set_sources(m_pool_watcher->get_images());
@@ -436,6 +445,7 @@ void Replayer::print_status(Formatter *f, stringstream *ss)
 
   if (f) {
     f->open_object_section("replayer_status");
+    f->dump_string("pool", m_local_io_ctx.get_pool_name());
     f->dump_stream("peer") << m_peer;
     f->open_array_section("image_replayers");
   };
diff --git a/src/tools/rbd_mirror/Replayer.h b/src/tools/rbd_mirror/Replayer.h
index 44b49e7..cd8efa8 100644
--- a/src/tools/rbd_mirror/Replayer.h
+++ b/src/tools/rbd_mirror/Replayer.h
@@ -85,7 +85,9 @@ private:
   std::map<std::string, std::unique_ptr<ImageReplayer<> > > m_image_replayers;
   std::unique_ptr<MirrorStatusWatchCtx> m_status_watcher;
 
+  std::string m_asok_hook_name;
   ReplayerAdminSocketHook *m_asok_hook;
+
   struct InitImageInfo {
     std::string global_id;
     std::string id;
